### PR TITLE
fix(tests): use riot lockfiles

### PR DESCRIPTION
**Pull request by emmettbutler**
This change makes the Riot test suite build its virtual environments using fully-specified dependency tree files that live in source control.

## How to Review

1. Review https://github.com/DataDog/dd-trace-py/pull/5227 and leave your comments on this change, or use [this filtered diff view](https://github.com/DataDog/dd-trace-py/pull/5246/files?file-filters%5B%5D=.py&file-filters%5B%5D=.rst&file-filters%5B%5D=.yml&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&show-deleted-files=true&show-viewed-files=true)

## Why?

The computation involved in building the tree of dependencies for a given test virtual environment is not small. By performing that work only when the tree changes rather than on every test run, we make our CI build time shorter.

Having the entire dependency tree in source control allows developers to unambiguously understand which versions of nth-order dependencies are in use in a given test environment without needing to run the tests locally, which can make debugging tests easier.

Using `riot run -c` to automatically rebuild these requirements files when necessary facilitates future changes like [having a bot rebuild and PR them periodically](https://github.com/DataDog/dd-trace-py/tree/emmett.butler/latest-automation).

## How?

By depending on riot==0.17.1, Riot runs on this branch look for requirements lockfiles and create them if they don't exist. That combined with the fact that the repo now contains a lockfile for each venv means that there are no changes necessary to the testing logic itself.

This change adds `scripts/compile-and-prune-test-requirements`. That script builds requirements.txt lockfiles for all Riot hashes in the current working tree that don't already have one in `.riot/requirements`. It also deletes such files for Riot hashes that don't exist. In a future change, this script could run under automation, commit its changes, and open pull requests.

## Notes

**If you're running tests locally and making changes to the riotfile, you'll need to run `riot` with the `-c` flag to ensure that those changes are reflected in the new dependency trees. You should commit these changes to dependency tree lockfiles and include them in pull requests that change the riotfile.**

This change removes the list of latest tested versions in `dependencies.py` as well as the logic that parses that list and uses it in tests. The finding of the latest version is now handled by Riot, which runs `pip-compile` internally.

Most of the changes to the riotfile in this diff are either:

* workarounds for the fact that versions of `pip-compile` prior to 3.7 don't do "[backtracking](https://pip.pypa.io/en/stable/topics/dependency-resolution/#backtracking)" during dependency resolution
* pins to the same versions as in the deleted `dependencies.py` file due to compile or test failures with newer versions

## Risk

The main risk I can think of is that this adds some confusion for contributors about how to get riot to notice their venv changes. I've updated the contributor docs accordingly.

## Checklist

- [x] Finish riot pull request https://github.com/DataDog/riot/pull/196
- [x] Finish the next riot pull request https://github.com/DataDog/riot/pull/202
- [x] Generate all requirements lockfiles
- [x] Fix riotfile to make tests pass on this PR
- [x] Move automation changes to separate PR
- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -407,6 +407,14 @@ jobs:
       - run_test:
           pattern: 'opentracer'
 
+  opentelemetry:
```
Comment (ZStriker19) [inline .circleci/config.yml:410]: This is part of adding the lockfiles? If so, what were we doing before for Otel testing? I think I'm probably missing something?

Comment (emmettbutler) [reply .circleci/config.yml:410]: This is a merge screwup. I'll fix the diff. The only parts to review in this PR are the gitignore and `.riot/requirements`.

**Review by brettlangdon — COMMENTED** [review]
I think we can also remove `./scripts/manage_dependencies.py`, right?

File: riotfile.py [hunk 2]
```diff
@@ -162,14 +148,14 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
-                "types-setuptools": latest,
-                "types-six": latest,
+                "mypy": "==0.991",
+                "envier": "==0.4.0",
+                "types-attrs": "==19.1.0",
+                "types-docutils": "==0.19.1.1",
+                "types-protobuf": "==3.20.4.5",
+                "types-PyYAML": "==6.0.12.2",
+                "types-setuptools": "==65.6.0.0",
+                "types-six": "==1.16.21.4",
```
Comment (brettlangdon) [inline riotfile.py:158]: we don't want any of these as "latest"?

File: tests/integration/test_debug.py [hunk 3]
```diff
@@ -266,7 +266,7 @@ def test_tracer_info_level_log(self):
         assert mock_logger.mock_calls == []
 
 
-@pytest.mark.subprocess(ddtrace_run=True)
+@pytest.mark.subprocess(ddtrace_run=True, err=None)
```
Comment (brettlangdon) [inline tests/integration/test_debug.py:269]: unrelated?

File: riotfile.py [hunk 4]
```diff
@@ -380,25 +366,38 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
                     pkgs={
                         "gevent": "~=1.3.0",
                         "greenlet": "~=1.0",
+                        "requests": "==2.20.0",
+                        "opensearch-py": "==1.0.0",
```
Comment (brettlangdon) [inline riotfile.py:370]: should we have comments for any of the places where we have exceptions?

it is unclear to me why we need to pin these here instead of relying on `latest` + `pip-compile` results?

File: riotfile.py (hunk 2 above)
Comment (emmettbutler) [reply riotfile.py:158]: I kept them at their previously-pinned versions from dependencies.py. We can make them track `latest` in a subsequent pull request.

Comment (brettlangdon) [reply riotfile.py:158]: ok, I am ok with updating them to latest in follow-up PR.

probably safer to pin now.

**Review by majorgreys — APPROVED** [review]
🎉

**Comment by brettlangdon** [comment]
Error from `stdlib` tests:

https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/31125/workflows/d59d5c65-20d2-4858-b851-17c9ec6263bb/jobs/2107199?invite=true#step-105-2586

```
Traceback (most recent call last):
  File "/root/project/.riot/venv_py3713_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451/bin/pytest", line 5, in <module>
    from pytest import console_main
  File "/root/project/.riot/venv_py3713_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451/lib/python3.7/site-packages/pytest/__init__.py", line 5, in <module>
    from _pytest._code import ExceptionInfo
  File "/root/project/.riot/venv_py3713_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451/lib/python3.7/site-packages/_pytest/_code/__init__.py", line 2, in <module>
    from .code import Code
  File "/root/project/.riot/venv_py3713_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451/lib/python3.7/site-packages/_pytest/_code/code.py", line 36, in <module>
    import pluggy
  File "/root/project/.riot/venv_py3713_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451/lib/python3.7/site-packages/pluggy/__init__.py", line 16, in <module>
    from ._manager import PluginManager, PluginValidationError
  File "/root/project/.riot/venv_py3713_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451/lib/python3.7/site-packages/pluggy/_manager.py", line 12, in <module>
    import importlib_metadata
ModuleNotFoundError: No module named 'importlib_metadata'
```


The profile tests failures seem flaky... ?

**Comment by emmettbutler** [comment]
@brettlangdon yeah, they are. I think that might be true of current 1.x as well.

**Review by brettlangdon — APPROVED** [review]

File: riotfile.py [hunk 5]
```diff
@@ -380,25 +366,38 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
+                ),
+                Venv(
+                    pys=select_pys(min_version="3.5", max_version="3.6"),
+                    pkgs={
+                        "gevent": "~=1.3.0",
+                        "greenlet": "~=1.0",
+                        "elasticsearch": "==6.3.1",
+                        "pynamodb": "==3.3.1",
+                        "requests": "==2.22.0",
+                        "six": "==1.12.0",
```
Comment (Yun-Kim) [inline riotfile.py:382]: Was six always needed as a requirement for gevent tests?

File: riotfile.py [hunk 6]
```diff
@@ -320,11 +301,19 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
                         Venv(pys=select_pys(max_version="3.5")),
-                        # DEV: attrs marked Python 3.6 as deprecated in 22.2.0,
-                        #      this logs a warning and causes these tests to fail
-                        # https://www.attrs.org/en/22.2.0/changelog.html#id1
-                        Venv(pys=["3.6"], pkgs={"attrs": "<22.2.0"}),
-                        Venv(pys=select_pys(min_version="3.7")),
+                        Venv(
+                            pkgs={
+                                "six": "==1.12.0",
+                            },
```
Comment (Yun-Kim) [inline riotfile.py:307]: was six always required for integration tests?

**Review by Yun-Kim — COMMENTED** [review]
Read through the changes to the riotfile and noticed some new dependencies that weren't present before. Could you clarify why they were added? Thanks!

File: riotfile.py [hunk 7]
```diff
@@ -2088,20 +2164,23 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
                     pys="2.7",
                     pkgs={
                         "snowflake-connector-python": "~=2.1.0",
+                        "pyOpenSSL": "~=19.1",
```
[the remark this answers is not in the document]
Comment (emmettbutler) [reply riotfile.py:2167]: @Yun-Kim sorry I missed your questions until now. These new dependencies were added as workarounds for the fact that versions of pip-compile prior to 3.7 don't do "[backtracking](https://pip.pypa.io/en/stable/topics/dependency-resolution/#backtracking)" during dependency resolution.
